overlord/devicestate/devicestate_systems_test.go

```diff
@@ -1236,6 +1240,255 @@ func (s *deviceMgrSystemsCreateSuite) TestDeviceManagerCreateRecoverySystemHappy
+	c.Check(s.bootloader.SetBootVarsCalls, Equals, 0)
+	c.Check(filepath.Join(boot.InitramfsUbuntuSeedDir, "systems", "1234", "snapd-new-file-log"), testutil.FileAbsent)
+}
+
+func (s *deviceMgrSystemsCreateSuite) TestDeviceManagerCreateRecoverySystemRemodelDownloadingMissingSnap(c *C) {
+	devicestate.SetBootOkRan(s.mgr, true)
+
+	fooSnap := snaptest.MakeTestSnapWithFiles(c, "name: foo\nversion: 1.0\nbase: core20", nil)
+	s.state.Lock()
+	defer s.state.Unlock()
+	// fake downloads are a nop
+	tSnapsup1 := s.state.NewTask("fake-download", "dummy task carrying snap setup")
+	// both snaps are asserted
+	snapsupFoo := snapstate.SnapSetup{
+		SideInfo: &snap.SideInfo{RealName: "foo", SnapID: s.ss.AssertedSnapID("foo"), Revision: snap.R(99)},
+		SnapPath: fooSnap,
+	}
+	tSnapsup1.Set("snap-setup", snapsupFoo)
+
+	tss, err := devicestate.CreateRecoverySystemTasks(s.state, "1234missingdownload", []string{tSnapsup1.ID()})
+	c.Assert(err, IsNil)
+	tsks := tss.Tasks()
+	c.Check(tsks, HasLen, 2)
+	tskCreate := tsks[0]
+	tskFinalize := tsks[1]
+	c.Assert(tskCreate.Summary(), Matches, `Create recovery system with label "1234missingdownload"`)
+	c.Check(tskFinalize.Summary(), Matches, `Finalize recovery system with label "1234missingdownload"`)
+	var systemSetupData map[string]interface{}
+	err = tskCreate.Get("recovery-system-setup", &systemSetupData)
+	c.Assert(err, IsNil)
+	c.Assert(systemSetupData, DeepEquals, map[string]interface{}{
+		"label":            "1234missingdownload",
+		"directory":        filepath.Join(boot.InitramfsUbuntuSeedDir, "systems/1234missingdownload"),
+		"snap-setup-tasks": []interface{}{tSnapsup1.ID()},
+	})
+	tss.WaitFor(tSnapsup1)
+	// add the dummy task to the change
+	chg := s.state.NewChange("create-recovery-system", "create recovery system")
+	chg.AddTask(tSnapsup1)
+	chg.AddAll(tss)
+
+	// downloads are only accepted if the tasks are executed as part of
+	// remodel, so procure a new model
+	newModel := s.brands.Model("canonical", "pc-20", map[string]interface{}{
+		"architecture": "amd64",
+		// UC20
+		"grade": "dangerous",
+		"base":  "core20",
+		"snaps": []interface{}{
+			map[string]interface{}{
+				"name":            "pc-kernel",
+				"id":              s.ss.AssertedSnapID("pc-kernel"),
+				"type":            "kernel",
+				"default-channel": "20",
+			},
+			map[string]interface{}{
+				"name":            "pc",
+				"id":              s.ss.AssertedSnapID("pc"),
+				"type":            "gadget",
+				"default-channel": "20",
+			},
+			// we have a download task for snap foo, but not for bar
+			map[string]interface{}{
+				"name":     "bar",
+				"presence": "required",
+			},
+		},
+		"revision": "2",
+	})
+	chg.Set("new-model", string(asserts.Encode(newModel)))
+
+	s.mockStandardSnapsModeenvAndBootloaderState(c)
+
+	s.state.Unlock()
+	s.settle(c)
+	s.state.Lock()
+
+	c.Assert(chg.Err(), ErrorMatches, `(?s).*cannot create a recovery system.*internal error: non-essential but "required" snap "bar" not present.`)
```

Comment:
I didn't notice when looking at the original code,  the quotes around required are a bit strange

Reply:
Are they? http://www.unnecessaryquotes.com/ ;-)